File: Parrot.java
/**
 * A Parrot object class to simulate pet parrot attributes
 *
 * */
public class Parrot {
    /**
     * Name of the Parrot pet
     * */
    private String name;
    /**
     * Health of Parrot measured in ints, range 0- 5. If he has more than 3.0 kgs in his stomach his health goes down
     * */
    private int health;
    /**
     * Crumbs to feed parrot in kg
     * */
    public double crumbs = 0.0;
    /**
     * Whether parrot is tamed
     * */
    private boolean isTamed;
    /**
     * Is parrot alive or not
     * */
    private boolean isAlive;
    /**
     * Is parrot sitting or not
     */
    private boolean isSitting;

    /**
     * Is parrot flying or not
     */
    private boolean isFlying;

    /**
     * This uses math.random to assign true or false to tame.
     * @return boolean to let someone know if a parrot is tamed or not
     */
    public boolean tameRandomly(){
        /* Will return true if the random number is between 0 and 0.5, false otherwise */
        boolean tame = (Math.random() < 0.50);
        return  tame;
    }

    /**
     *
     * @return boolean to let consumer know whether a parrot is alive or not.
     */
    public boolean isAlive() {
        return isAlive;
    }

    /**
     * This sets the parrot's alive status to true or false
     * @param alive
     */
    public void setAlive(boolean alive) {
        isAlive = alive;
    }

    /**
     * This sets the parrot's tamed status to true or false
     * @param tamed
     */
    public void setTamed(boolean tamed) {
        isTamed = tamed;
    }

    /**
     * This helps someone know if a parrot is tamed or not
     * @return boolean for whether parrot is tamed or not
     */
    public boolean getTamedStatus(){
        return this.isTamed;
    }

    /**
     * This returns the name of the parrot
     * @return string of name of parrot
     */
    public String getName() {
        return name;
    }

    /**
     * This sets the name of a parrot
     * @param name is used to set name of parrot
     */
    public void setName(String name) {
        this.name = name;
    }

    /**
     * Returns health of a parrot measured in int units
     * @return health value
     */
    public int getHealth() {
        return this.health;
    }

    /**
     * Lowers health of parrot based on int health amount
     * @param health is how many health units to decrease for parrot
     */
    public void lowerHealth(int health) {
        this.health = this.health - health;
    }

    /**
     * Increases health of parrod based on int health amount
     * @param health is how many health units to increase for parrot
     */
    public void increaseHealth(int health) {
        this.health = this.health + health;
    }

    /**
     * Constructor for Parrot with default  name "Polly"
     */
    public Parrot(){
        this.name = "Polly";
        this.crumbs = 0.1;
        this.health = 3;
        this.isTamed = false;
        this.isAlive = true;
        this.isSitting = false;
        this.isFlying = true;

    }
    /**
     * Creates a customized Parrot object
     *
     * @param name will be the name of the parrot
     */
    public Parrot(String name){
        this.name = name;
        this.health = 3;
        this.isTamed = false;
        this.isAlive = true;
        this.crumbs = 0.1;
        this.isSitting = false;
        this.isFlying = true;

    }

    /**
     * Returns a short summary of a parrot's current state on Name, health, istamed, isAlive, isFlying, isSitting
     * and how many crumbs it has in its stomach
     * @return string report
     */
    @Override
    public String toString() {
        return "\nParrot Stats"+"\n--------------"+ "\nName:"+
                this.name + "\nHealth: "+
                this.health +"\nisTamed: "+
                this.isTamed +"\nisAlive:  "+ this.isAlive +
                "\nIs Flying: "+
                this.isFlying + "\nIs Sitting: "+
                this.isSitting + "\nHas this many crumbs in stomach: "+ this.crumbs +"\n";
    }

    /**
     * This is a hardcoded overloaded toString method because I wanted all the parrot stats to display horizontally
     * @param p is the first parrot object
     * @param p1 is the second parrot object
     * @param p2 is the third parrot object
     * @return String which is a report on all the parrot attributes such as name, health, isTamed, isAlive, isFlying, is
     * Sitting, has crumbs in stomach
     */
    public String toString(Parrot p, Parrot p1, Parrot p2) {
        return "Parrot Stats"+"\n----------------------------------------------------------------------"+ "\nName:"+
                p.getName() + "\t\t\t\t\t"+ "Name:"+p1.getName() + "\t\t\t\t"+ "Name:"+ p2.getName() +
                "\nHealth: "+
                p.getHealth() + "\t\t\t\t\t"+ "Health: "+ p1.getHealth() + "\t\t\t\t"+ "Health: "+ p2.getHealth() +"\t\t"	+"\nisTamed: "+
                p.getTamedStatus() + "\t\t\t\t"+ "isTamed: "+ p1.getTamedStatus() + "\t\t\t"+ "isTamed: "+ p2.getTamedStatus() +"\nisAlive:  "+
                p.isAlive() + "\t\t\t\t"+  "isAlive: "+ p1.isAlive() + "\t\t\t"+  "isAlive: "+ p2.isAlive() +
                "\nIs Flying: "+
                p.isFlying() + "\t\t\t\t"+ "isFlying: "+ p1.isFlying() + "\t\t\t"+ "isFlying: "+ p2.isFlying() + "\nIs Sitting: "+
                p.isSitting() + "\t\t\t"+ "isSitting: "+ p1.isSitting() + "\t\t"+ "isSitting: "+ p2.isSitting() +
                "\ncrumbs in stomach: "+ p.getCrumbs() +"kg "+ "\t"+  "crumbs in stomach: "+ p1.getCrumbs() +"kg "+
                "crumbs in stomach: "+  p2.getCrumbs()+"kg ";
    }

    /**
     * This sets the amount of crumbs in a parrot's stomach by increasing by 1
     */
    public void setCrumbs(){
        this.crumbs += 1;
    }

    /**
     * This returns how many crumbs are in a parrot's stomach
     * @return the crumb amount in double format
     */
    public double getCrumbs(){
        return this.crumbs;
    }

    /**
     * This sets the parrot's status of sitting to be true or false
     * @param sitting is a boolean to toggle a parrot sitting
     */
    public void setSitting(Boolean sitting){
        this.isSitting = sitting;
    }

    /**
     * This returns the boolean for whether a parrot is flying
     * @return boolean for if parrot is flying
     */
    public boolean isFlying(){
        return this.isFlying;
    }
    /**
     * This sets the parrot's status of flying to be true or false
     * @param flying is a boolean to toggle a parrot flying
     */
    public void setFlying(Boolean flying){
        this.isFlying = flying;
    }

    /**
     * This returns whether a parrot is sitting or not
     * @return boolean for parrot's sitting status
     */
    public boolean isSitting(){
        return this.isSitting;
    }

}

File: parrot.rs
use rand::random;

#[derive(Debug, Clone)]
pub struct Parrot {
    name: String,
    health: i32,
    crumbs: f64,
    is_tamed: bool,
    is_alive: bool,
    is_sitting: bool,
    is_flying: bool,
}

impl Parrot {
    // Default constructor
    pub fn new() -> Self {
        Self {
            name: "Polly".to_string(),
            health: 3,
            crumbs: 0.1,
            is_tamed: false,
            is_alive: true,
            is_sitting: false,
            is_flying: true,
        }
    }

    // Constructor with custom name
    pub fn with_name(name: &str) -> Self {
        Self {
            name: name.to_string(),
            health: 3,
            crumbs: 0.1,
            is_tamed: false,
            is_alive: true,
            is_sitting: false,
            is_flying: true,
        }
    }

    pub fn tame_randomly(&mut self) -> bool {
        let tamed = random::<f64>() < 0.5;
        self.is_tamed = tamed;
        tamed
    }

    pub fn is_alive(&self) -> bool {
        self.is_alive
    }

    pub fn set_alive(&mut self, alive: bool) {
        self.is_alive = alive;
    }

    pub fn set_tamed(&mut self, tamed: bool) {
        self.is_tamed = tamed;
    }

    pub fn get_tamed_status(&self) -> bool {
        self.is_tamed
    }

    pub fn get_name(&self) -> &str {
        &self.name
    }

    pub fn set_name(&mut self, name: &str) {
        self.name = name.to_string();
    }

    pub fn get_health(&self) -> i32 {
        self.health
    }

    pub fn lower_health(&mut self, amount: i32) {
        self.health -= amount;
    }

    pub fn increase_health(&mut self, amount: i32) {
        self.health += amount;
    }

    pub fn set_crumbs(&mut self) {
        self.crumbs += 1.0;
    }

    pub fn get_crumbs(&self) -> f64 {
        self.crumbs
    }

    pub fn set_sitting(&mut self, sitting: bool) {
        self.is_sitting = sitting;
    }

    pub fn is_flying(&self) -> bool {
        self.is_flying
    }

    pub fn set_flying(&mut self, flying: bool) {
        self.is_flying = flying;
    }

    pub fn is_sitting(&self) -> bool {
        self.is_sitting
    }

    pub fn to_string(&self) -> String {
        format!(
            "\nParrot Stats\n--------------\nName: {}\nHealth: {}\nisTamed: {}\nisAlive: {}\nIs Flying: {}\nIs Sitting: {}\nHas this many crumbs in stomach: {:.1}\n",
            self.name,
            self.health,
            self.is_tamed,
            self.is_alive,
            self.is_flying,
            self.is_sitting,
            self.crumbs
        )
    }

    pub fn to_string_multiple(p1: &Parrot, p2: &Parrot, p3: &Parrot) -> String {
        format!(
            "Parrot Stats\n----------------------------------------------------------------------\nName: {}\t\t\t\tName: {}\t\t\t\tName: {}\nHealth: {}\t\t\t\tHealth: {}\t\t\t\tHealth: {}\nisTamed: {}\t\t\t\tisTamed: {}\t\t\tisTamed: {}\nisAlive: {}\t\t\tisAlive: {}\t\t\tisAlive: {}\nIs Flying: {}\t\t\tisFlying: {}\t\t\tisFlying: {}\nIs Sitting: {}\t\tisSitting: {}\t\tisSitting: {}\ncrumbs in stomach: {:.1}kg\tcrumbs in stomach: {:.1}kg\tcrumbs in stomach: {:.1}kg",
            p1.name, p2.name, p3.name,
            p1.health, p2.health, p3.health,
            p1.is_tamed, p2.is_tamed, p3.is_tamed,
            p1.is_alive, p2.is_alive, p3.is_alive,
            p1.is_flying, p2.is_flying, p3.is_flying,
            p1.is_sitting, p2.is_sitting, p3.is_sitting,
            p1.crumbs, p2.crumbs, p3.crumbs
        )
    }
}
